src/domains/Action.ts

```diff
@@ -15,15 +15,13 @@
  */
 
 import { ActionType } from "./ActionType";
+import { nanoid } from 'nanoid';
 
 export class Action {
+    public id: string = nanoid(4);
     public type: ActionType;
-    /**
-     * output path from the previous action
-     */
-    public lastOutput: string;
-    /**
-     * the index of actions
-     */
-    public index: number;
+    // for serialization
+    public upstreamActionIds: string[] = [];
+    public downstreamIds: string[] = [];
+    public outputFilename: string;
```

Comment:
nit: wonder why this is not optional?

Reply:
I will check that later, the outputFilename may be a deprecate property. the previous version the action is used for running

Reply:
OK. I confirmed that `outputFilename` cannot be removed, but this is optional. the `outputFilename` is used to determine the `outputPath` of the vertex of the action. if `outputFilename`, vertex id will be used.

Here is two working VideoProcessor implementation code which use this property.

https://github.com/irohalab/mira-video-manager/blob/refactor-action/src/processors/LocalConvertProcessor.ts#L56

https://github.com/irohalab/mira-video-manager/blob/refactor-action/src/processors/LocalExtractProcessor.ts#L59

Reply:
I should add some comment to mark it optional and its usage.

Reply:
Besides, all actions will have an output for the final result or for the downstreams.